proposals/0000-midstream-bindists.md

```diff
@@ -0,0 +1,199 @@
+[GHCup is not an installer](https://hasufell.github.io/posts/2023-11-14-ghcup-is-not-an-installer.html) recently.
+
+## Problem Statement
+
+From the perspective of a GHCup developer, there are several issues with relying on upstream bindists.
+
+### Platform support
+
+GHC and other tools have in the past dropped support for certain platforms either entirely or requested
+the community to step up and do the work (e.g. on GHC CI).
+
+E.g. [GHC ARMv7 support was dropped silently without any call for help](https://gitlab.haskell.org/ghc/ghc/-/issues/21177#note_470440). Similarly, FreeBSD support just ceased to exist when the GHC FreeBSD CI stopped working. Later the community asked for a [revival](https://gitlab.haskell.org/groups/ghc/-/epics/5), but nothing signifcant has happened so far.
+GHCup still produces bindists from time to time for FreeBSD, but e.g. the [HLS release manager for 2.5.0.0 recently refused to add FreeBSD bindists](https://github.com/haskell/ghcup-metadata/pull/159).
+
+Similarly, stack used to have issues with Aarch64 bindists:
+
+* https://github.com/commercialhaskell/stack/issues/5709
+* https://github.com/commercialhaskell/stack/issues/5854
+* https://github.com/commercialhaskell/stack/issues/5540
+* https://github.com/commercialhaskell/stack/issues/5610
+* https://github.com/commercialhaskell/stack/issues/5619
+* https://github.com/commercialhaskell/stack/issues/6141
+* https://github.com/commercialhaskell/stack/issues/6142
+
+Recently, cabal-install had issues with i386 binaries and alpine, delaying a GHCup metadata PR:
+
+* https://github.com/haskell/ghcup-metadata/pull/127#issuecomment-1766020410
+
+These issues are frequent and so far the GHCup developers used to single handedly fix all those missing bindists manually
+and provide them here: https://downloads.haskell.org/~ghcup/unofficial-bindists/
+
+This is unfunded and significant work.
+
+### Bindist maintenance
+
+Sometimes, bindists are broken, e.g. for GHC there are a couple of instances:
+
+* 9.0.2 shipping without profiling info: https://gitlab.haskell.org/ghc/ghc/-/issues/21841
+* DESTDIR variable ignored by `make install`: https://gitlab.haskell.org/ghc/ghc/-/issues/19646
+
+Sometimes, bindists have been built for very old version of linux distros and won't run well on newer linux versions.
+This is currently a problem since Debian has removed ncurses5:
+
+* https://github.com/haskell/ghcup-hs/issues/902
+* https://answers.launchpad.net/ubuntu/+source/ncurses/+question/707838
+* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1025964
+
+For cabal, there has been the infamous hLock issue:
+
+* https://github.com/haskell/cabal/issues/7313
+* https://github.com/haskell/cabal/issues/7950
+
+This shows that bindists, for current and historical versions, need continuous maintenance. However, upstream developers
+so far have very rarely engaged in this type of maintenance work, pushing it down to GHCup. As an example, here are all the
```

Comment:
> Someone in the community is fixing the bindists but not pushing them upstream?

I tried to get i386 alpine bindists into GHC: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5213

After 2 years, it still hasn't happened yet (the issue got stuck while Ben was debugging some issues with statically linked GHCs, which I did not request at all).

----

Instead with the proposed framework, I believe that upstream developers would actually benefit from this, because:

- they don't have to build releases at all anymore
- if they still build releases, they can focus on smaller coverage (e.g. not build for every esoteric linux distro)
- test results for "unsupported" or tier3 platforms can flow back to upstream developers, allowing ad-hoc fixes even if there's no continuous GHC CI support for said platform

Reply:
> Additionally, fixing upstream bindists has a much bigger impact than just on GHCup, because upstream's bindists are used by other packagers, e.g. linux distributions, nixpkgs, etc.

FWIW I think all the packagers you mention build their own binary distribution, they don't repackage the upstream bindist.

Reply:
Yes I think you're right. Nixpkgs at least only uses them for bootstrapping. That's not an unimportant use-case, though, and you certainly have a problem if your bootstrapping bindists don't work.

Reply:
At that point it seems like we've removed all reason for upstream to bother producing bindists at all. Maybe we _should_ say that - it would remove some repeat labour. But if nothing else it's probably good for upstream to at least notice _some_ packaging problems quickly.

Reply:
Maybe, but that's up to upstream. We don't want to tell them what to do.